query-projector/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+            .iter()
+            .map(|ti| ti.lookup(&row))
+            .collect::<Result<Vec<TypedValue>>>()
+    }
+}
+
+impl Projector for TupleProjector {
+    fn project<'stmt>(&self, mut rows: Rows<'stmt>) -> Result<QueryResults> {
+        if let Some(r) = rows.next() {
+            let row = r?;
+            let bindings = self.collect_bindings(row)?;
+            Ok(QueryResults::Tuple(Some(bindings)))
+        } else {
+            Ok(QueryResults::Tuple(None))
+        }
+    }
+}
+
+/// A rel projector produces a vector of vectors.
+/// Each inner vector is the same size, and sourced from the same columns.
+/// One inner vector is produced per `Row`.
+/// Each column in the inner vector is the result of taking one or two columns from
+/// the `Row`: one for the value and optionally one for the type tag.
+struct RelProjector {
+    len: usize,
+    templates: Vec<TypedIndex>,
+}
+
+impl RelProjector {
+    fn with_templates(len: usize, templates: Vec<TypedIndex>) -> RelProjector {
+        RelProjector {
+            len: len,
+            templates: templates,
+        }
+    }
+
+    fn collect_bindings<'a, 'stmt>(&self, row: Row<'a, 'stmt>) -> Result<Vec<TypedValue>> {
+        assert_eq!(row.column_count(), self.len as i32);
+        self.templates
+            .iter()
+            .map(|ti| ti.lookup(&row))
+            .collect::<Result<Vec<TypedValue>>>()
+    }
+}
+
+impl Projector for RelProjector {
+    fn project<'stmt>(&self, mut rows: Rows<'stmt>) -> Result<QueryResults> {
+        let mut out: Vec<Vec<TypedValue>> = vec![];
+        while let Some(r) = rows.next() {
+            let row = r?;
+            let bindings = self.collect_bindings(row)?;
+            out.push(bindings);
+        }
+        Ok(QueryResults::Rel(out))
+    }
+}
+
+/// A coll projector produces a vector of values.
+/// Each value is sourced from the same column.
+struct CollProjector {
+    template: TypedIndex,
+}
+
+impl CollProjector {
+    fn with_template(template: TypedIndex) -> CollProjector {
+        CollProjector {
+            template: template,
+        }
+    }
+}
+
+impl Projector for CollProjector {
+    fn project<'stmt>(&self, mut rows: Rows<'stmt>) -> Result<QueryResults> {
+        let mut out: Vec<TypedValue> = vec![];
+        while let Some(r) = rows.next() {
+            let row = r?;
+            let binding = self.template.lookup(&row)?;
+            out.push(binding);
+        }
+        Ok(QueryResults::Coll(out))
+    }
+}
+
+/// Convenience function for grabbing a single-element projection.
+fn one_typed_index(element: &Element, query: &AlgebraicQuery, mut cols: &mut Vec<ProjectedColumn>) -> TypedIndex {
+    let mut templates = project_elements(iter::once(element), 0, query, &mut cols);
+    assert_eq!(1, templates.len());
+    templates.pop().expect("Couldn't pop a vec of len 1!")
+}
+
+/// Combines the two things you need to turn a query into SQL and turn its results into
+/// `QueryResults`.
```

Comment:
The `README.md` has a nice description of the two things.  It's worth repeating.

nit: consider s/projector/datalog_projector/ or similar to really drive the "two sides, one coin" point.  I found the two projections/projectors a little tricky to keep track of why reading this, and had to refer to the high-level note when I read this.